validator nonces are now supplied by the user

CluEleSsUK opened the pull request:
- operator must now pass their validator nonce to the key signing command
- operators must add their validator nonce when uploading their key
- users must supply validator nonces when starting the DKG
- documentation updated to reflect the changes

CluEleSsUK committed:
validator nonces are now supplied by the user

- operator must now pass their validator nonce to the key signing
  command
- operators must add their validator nonce when uploading their key
- users must supply validator nonces when starting the DKG
- documentation updated to reflect the changes

AnomalRoil commented on cli/sign.go:
```diff
@@ -96,6 +107,27 @@ func Sign(operators []string, depositData []byte, log shared.QuietLogger) ([]api
+
+	validatorNonce, err := strconv.Atoi(parts[0])
+	if err != nil {
+		return 0, "", errors.New("validatorNonce for the operator must be a number")
```
any reason not to wrap the existing `err`?
```suggestion
		return 0, "", fmt.Errorf("validatorNonce for the operator must be a number: %w", err)
```

CluEleSsUK replied:
it bubbles through to the CLI, so I didn't wanna expose anything to mad to the user

AnomalRoil commented on integration/integration_test.go:
```diff
@@ -23,7 +24,7 @@ func TestSuccessfulSigning(t *testing.T) {
 
 	operators := fmap(ports, func(o uint) string {
-		return fmt.Sprintf("http://localhost:%d", o)
+		return fmt.Sprintf("%d,http://localhost:%d", o, o)
```
wait, is the port the same as the index?

CluEleSsUK replied:
for the test, yes - the index in this case isn't the DKG, but rather a nonce issued by the SSV registration smart contract

AnomalRoil commented on nodes/README.md:
```diff
@@ -2,6 +2,7 @@
 This module contains a single file containing all the SSV nodes that support the DKG sidecar.
 
-Node operators should use the `sign` functionality of the sidecar CLI to sign their public key, and raise a pull request with the output of the sign command appended to the `operators.json` file.
+Node operators should use the `sign` functionality of the sidecar CLI with their registered SSV validator nonce to sign their public key, and raise a pull request with the output of the sign command appended to the `operators.json` file.
```
Maybe provide example command? 
This is hard to comprehend for me, as a non-SSV validator, but maybe they are used to it? As you want.

CluEleSsUK replied:
added a link to the docs for it

AnomalRoil commented on nodes/README.md:
```diff
@@ -2,6 +2,7 @@
-Node operators should use the `sign` functionality of the sidecar CLI to sign their public key, and raise a pull request with the output of the sign command appended to the `operators.json` file.
+Node operators should use the `sign` functionality of the sidecar CLI with their registered SSV validator nonce to sign their public key, and raise a pull request with the output of the sign command appended to the `operators.json` file.
 
+Triple check your validator nonce - if you use an incorrect one, you will be unable to receive recompensation for validator work.
```
```suggestion
Triple check your validator nonce - if you use an incorrect one, you will be unable to receive reward for validator work.
```

AnomalRoil commented on sidecar/README.md:
```diff
@@ -17,8 +17,9 @@ $ ssv-sidecar key create ~/.ssv
-$ ssv-sidecar key sign --directory ~/.ssv --url https://example.org | jq
+$ ssv-sidecar key sign --validator-nonce 2 --directory ~/.ssv --url https://example.org | jq
 {
+  "validator_nonce": 2,
```
are validator nonces usually of this size?

CluEleSsUK replied:
apparently it's just a monotonically increasing number in the SSV smart contract (and there is no validator with `0`)

AnomalRoil commented on sidecar/internal/cmd/root.go:
```diff
@@ -2,10 +2,12 @@ package cmd
 )
 
 var DirectoryFlag string
+var ValidatorNonceFlag uint32
```
Why not put it in the `cmd/key.go` file since that's where you declare it's used for a flag?

AnomalRoil commented on sidecar/keys.go:
```diff
@@ -38,6 +42,8 @@ func SignKey(url string, keyPath string) ([]byte, error) {
 		return nil, fmt.Errorf("failed to sign address: %w", err)
 	}
 
+	identity.ValidatorNonce = validatorNonce
```
So the validatorNonce isn't `SelfSign`ed, I think maybe it should be passed to `keypair.SelfSign` and be included in the signature, no, to make sure to authenticate it?

CluEleSsUK replied:
done

AnomalRoil replied:
since  `keypair.SelfSign` returns the identity, I think this step should be done inside of it, no?

CluEleSsUK replied:
oops it already is, good catch

AnomalRoil commented on cli/internal/cmd/operators_list.go:
```diff
@@ -53,22 +54,22 @@ func listOperators(_ *cobra.Command, _ []string) {
 
 	// verify the signatures of all the operators
-	var addresses []string
+	var operatorIdentities []crypto.Identity
```
Given you know the size of the `operators` array, this is probably the kind of cases where you either want to:
- use a fixed size array, but no one does that usually
- use a slice made with `make([]crypto.Identity,0, len(operators))` so that it's capacity is already correct, to avoid `append` having to allocate a bigger array in each iteration

AnomalRoil commented on cli/internal/cmd/operators_list.go:
```diff
@@ -107,28 +108,36 @@ func readSourceUrl(url string) []crypto.Identity {
-	var failure []string
-	for _, address := range addresses {
-		res, err := http.Get(fmt.Sprintf("%s/health", address))
+	var success []crypto.Identity
+	var failure []crypto.Identity
+	for _, o := range operators {
```
same comment about `append` and using make() to set capacity.

CluEleSsUK committed:
validator nonces are now provided by the user

- keys are signed by providing a nonce
- key verifier tool now requires the nonce
- sign APIs updated to POST to take the nonce
- nonce removed from the sidecar API (as the user provides it)

CluEleSsUK committed:
updated sample operators

AnomalRoil reviewed (commented):
I would have expected the ValidatorNonce to be incremented somewhere, no? I don't see that? Or is it fixed forever?

Also if the validator nonce is part of the signature, this means signatures are changing for each message I just realized, which might not be what we want?

We might want to discuss how the `SelfSign` signature and how the `ValidatorNonce` are meant to be used to avoid misunderstandings.

CluEleSsUK commented:
ValidatorNonce is SSV parlance for a OperatorID - the SSV smart contract issues the operator a number when they sign up to become a validator node. It never changes, but is unique between operators

CluEleSsUK commented:
SelfSign is used to add the validator public key to the repo to provide attribution. There’s some weirdness that the user provides the nonce when they start the DKG (rather than the operator starting the sidecar with the nonce). I queried this with SSV, but they suggested it’s up to the user to provide the right one. There’s a race condition wrt to starting an SSV node, as you can start an SSV node before you get a validator nonce; perhaps we could mandate one is required to run the DKG sidecar?

CluEleSsUK committed:
remove re-setting of validator nonce in signing

AnomalRoil commented on shared/crypto/keys.go:
```diff
@@ -29,29 +33,59 @@ type Keypair struct {
 	return suite.Verify(m, i.Public, i.Signature)
 }
+
+func messageFor(publicKey []byte, address string, validatorNonce uint32) ([]byte, error) {
+	buf := new(bytes.Buffer)
+
+	if err := binary.Write(buf, binary.BigEndian, publicKey); err != nil {
+		return nil, err
+	}
+
+	if err := binary.Write(buf, binary.BigEndian, []byte(address)); err != nil {
+		return nil, err
+	}
+
+	if err := binary.Write(buf, binary.BigEndian, validatorNonce); err != nil {
+		return nil, err
+	}
+
+	out := sha256.New()
+	if _, err := out.Write(buf.Bytes()); err != nil {
+		return nil, err
+	}
+	return out.Sum(nil), nil
```
This looks more like a "digestFor" or "hashFrom" function.
Also, it might be a good idea to have some sort of domain separation tags to avoid risks of clashes, or is this formatting dictated by the SSV side of things?

AnomalRoil replied:
That being said, I'd be hard pressed to find a collision, but since these are user-provided values...

CluEleSsUK replied:
yep that's a good idea - will add it.

AnomalRoil reviewed (approved):
lgtm I guess.

Have you considered adding `golangci-lint` as a github action maybe?

CluEleSsUK committed:
added a DST to the key message digest

CluEleSsUK committed:
updated keys with dst

CluEleSsUK committed:
fixed some tests
